pkg/fakelinstor/fakelinstor.go

```diff
@@ -0,0 +1,234 @@
+	defer r.Body.Close() //nolint:errcheck
+
+	var node lapi.Node
+	err := json.NewDecoder(r.Body).Decode(&node)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding node: %w", err)
+	}
+
+	if len(node.NetInterfaces) > 0 && node.Type == linstor.ValNodeTypeStlt {
+		node.ConnectionStatus = "ONLINE"
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	for i := range f.nodes {
+		if f.nodes[i].Name == node.Name {
+			return nil, fmt.Errorf("node '%s' already exists", node.Name)
+		}
+	}
+	f.nodes = append(f.nodes, node)
+
+	return nil, nil
+}
+
+func (f *fakeLinstor) getNode(r *http.Request) (any, error) {
+	node := r.PathValue("node")
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	for i := range f.nodes {
+		if f.nodes[i].Name == node {
+			return f.nodes[i], nil
+		}
+	}
+
+	return nil, lapi.NotFoundError
+}
+
+func (f *fakeLinstor) deleteNode(r *http.Request) (any, error) {
+	node := r.PathValue("node")
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	for i := range f.resources {
+		if f.resources[i].NodeName == node {
+			return nil, fmt.Errorf("cannot delete node '%s', it still contains resource '%s'", node, f.resources[i].Name)
+		}
+	}
+
+	f.nodes = slices.DeleteFunc(f.nodes, func(n lapi.Node) bool {
+		return n.Name == node
+	})
+
+	return nil, nil
+}
+
+func (f *fakeLinstor) createResourceDefinition(r *http.Request) (any, error) {
+	defer r.Body.Close() //nolint:errcheck
+
+	var rd lapi.ResourceDefinitionCreate
+	err := json.NewDecoder(r.Body).Decode(&rd)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding resource definition: %w", err)
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if slices.ContainsFunc(f.resourceDefinitions, func(exist lapi.ResourceDefinition) bool {
+		return exist.Name == rd.ResourceDefinition.Name
+	}) {
+		return nil, fmt.Errorf("resource definition '%s' already exists", rd.ResourceDefinition.Name)
+	}
+
+	f.resourceDefinitions = append(f.resourceDefinitions, rd.ResourceDefinition)
+	return nil, nil
+}
+
+func (f *fakeLinstor) createResource(r *http.Request) (any, error) {
+	defer r.Body.Close() //nolint:errcheck
+
+	var rc lapi.ResourceCreate
+	err := json.NewDecoder(r.Body).Decode(&rc)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding resource create: %w", err)
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	rd := r.PathValue("rd")
+	node := r.PathValue("node")
+	rc.Resource.Name = rd
+	rc.Resource.NodeName = node
+
+	if !slices.ContainsFunc(f.resourceDefinitions, func(exist lapi.ResourceDefinition) bool {
+		return exist.Name == rd
+	}) {
+		return nil, lapi.NotFoundError
+	}
+
+	if !slices.ContainsFunc(f.nodes, func(exist lapi.Node) bool {
+		return exist.Name == node
+	}) {
+		return nil, lapi.NotFoundError
+	}
+
+	if !slices.ContainsFunc(f.resources, func(exist lapi.ResourceWithVolumes) bool {
+		return exist.Name == rd && exist.NodeName == node
+	}) {
+		f.resources = append(f.resources, lapi.ResourceWithVolumes{
+			Resource: rc.Resource,
+		})
+	}
+
+	return nil, nil
```

Comment:
I guess it is deliberate and consistent with the real Controller to not return an error when the resource already exists?

Reply:
Yes, a resource create on an existing node is a no-op in LINSTOR.